core/src/main/java/com/digitalpebble/storm/crawler/Metadata.java

```diff
@@ -0,0 +1,145 @@
+/**
+ * Licensed to DigitalPebble Ltd under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * DigitalPebble licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.digitalpebble.storm.crawler;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.apache.commons.lang.StringUtils;
+
+/** Wrapper around Map <String,String[]> **/
+
+public class Metadata {
+
+    // TODO customize the behaviour of Kryo via annotations
+    // @BindMap(valueSerializer = IntArraySerializer.class, keySerializer =
+    // StringSerializer.class, valueClass = String[].class, keyClass =
+    // String.class, keysCanBeNull = false)
+    private Map<String, String[]> md;
+
+    public final static Metadata empty = new Metadata(
+            Collections.<String, String[]> emptyMap());
+
+    public Metadata() {
+        md = new HashMap<String, String[]>();
+    }
+
+    /**
+     * Wraps an existing HashMap into a Metadata object - does not clone the
+     * content
+     **/
+    public Metadata(Map<String, String[]> metadata) {
+        if (metadata == null)
+            throw new NullPointerException();
+        md = metadata;
+    }
+
+    public Map<String, String[]> getMap() {
+        return md;
+    }
+
+    /** @returns the first value for the key or null if it does not exist **/
+    public String getFirstValue(String key) {
+        String[] values = md.get(key);
+        if (values == null)
+            return null;
+        if (values.length == 0)
+            return null;
+        return values[0];
+    }
+
+    public String[] getValues(String key) {
+        String[] values = md.get(key);
+        if (values == null)
+            return null;
+        if (values.length == 0)
+            return null;
+        return values;
+    }
+
+    public void setValue(String key, String value) {
+        md.put(key, new String[] { value });
+    }
+
+    public void setValues(String key, String[] values) {
+        md.put(key, values);
```

Comment:
I can think of cases where we'd want to store only a key and not be interested in the values. let's leave it as is